querier.cc

```diff
@@ -152,11 +160,19 @@ static bool ranges_match(const schema& s, dht::partition_ranges_view original_ra
 }
 
 template <typename Querier>
-static can_use can_be_used_for_page(const Querier& q, const schema& s, const dht::partition_range& range, const query::partition_slice& slice) {
+static can_use can_be_used_for_page(const replica::database& db, Querier& q, const schema& s, const dht::partition_range& range, const query::partition_slice& slice, reader_concurrency_semaphore& current_sem) {
     if (s.version() != q.schema().version()) {
         return can_use::no_schema_version_mismatch;
     }
 
+    auto& querier_sem = q.permit().semaphore();
+    if (&querier_sem != &current_sem) {
+        if (db.is_user_semaphore(querier_sem) && db.is_user_semaphore(current_sem)) {
```

Comment:
If this is all we use the `database&` for, maybe we can just pass in a small function object: `std::function<bool(const reader_concurrency_semaphore&)> is_user_semaphore_func`. To avoid pulling in the huge `replica/database.hh` just for this.